numba/tests/test_tuples.py

```diff
@@ -609,6 +609,178 @@ def test_index(self):
         self.assertEqual(msg, str(raises.exception))
 
 
+class TestIntrinsics(TestCase):
+    def test_make_tuple_types(self):
+        from numba.targets.tupleobj import make_tuple
+    
+        @njit
+        def foo():
+            a = make_tuple(3, list(range(3)))
+            b = make_tuple(3, range(3))
+            c = make_tuple(3, zip(range(0, 3), range(3, 6)))
+            it = iter(range(3))
+            d = make_tuple(3, it)
+            arr = np.array(list(range(3)))
+            e = make_tuple(3, arr)
+            f = make_tuple(3, [(True, True), (True, False), (False, False)])
+            return a, b, c, d, e, f
+
+        a, b, c, d, e, f = foo()
+        expected1 = tuple(range(3))
+        self.assertEqual(a, expected1)
+        self.assertEqual(b, expected1)
+        self.assertEqual(c, tuple(zip(range(0, 3), range(3, 6))))
+        self.assertEqual(d, expected1)
+        self.assertEqual(e, expected1)
+        self.assertEqual(f, tuple([(True, True), (True, False), (False, False)]))
+
+    def test_make_tuple_repeated_call(self):
+        from numba.targets.tupleobj import make_tuple
+    
+        @njit
+        def foo():
+            it_a = list(range(3))
+            a1 = make_tuple(3, it_a)
+            a2 = make_tuple(3, it_a)
+            it_b = range(6)
+            b1 = make_tuple(3, it_b)
+            b2 = make_tuple(3, it_b)
+            it_c = zip(range(0, 6), range(6, 12))
+            c1 = make_tuple(3, it_c)
+            c2 = make_tuple(3, it_c)
+            it_d = iter(range(6))
+            d1 = make_tuple(3, it_d)
+            d2 = make_tuple(3, it_d)
+            arr = np.array(list(range(3)))
+            e1 = make_tuple(3, arr)
+            e2 = make_tuple(3, arr)
+            it_f = [(True, True), (True, False), (False, False), (False, True)]
+            f1 = make_tuple(2, it_f)
+            f2 = make_tuple(2, it_f)
```

Comment:
might be worth adding a typed-list and typed-dict to this.

Reply:
Done.